api/v1beta2/mysqlcluster_types.go

```diff
@@ -259,17 +266,104 @@ func (s MySQLClusterSpec) validateUpdate(old MySQLClusterSpec) field.ErrorList {
 			newSize := pvc.StorageSize()
 			oldSize := old.StorageSize()
 
-			if newSize.Cmp(oldSize) == -1 {
+			switch cmp := newSize.Cmp(oldSize); {
+			case cmp == -1:
 				p := p.Child("volumeClaimTemplates").Index(i).
 					Child("spec").Child("resources").Child("requests").Key("storage")
 				allErrs = append(allErrs, field.Forbidden(p, "storage size cannot be reduced"))
+			case cmp == 1:
+				isVolumeExpansion = true
+			case cmp == 0:
+				// noop
 			}
 		}
 	}
 
+	if isVolumeExpansion {
+		allErrs = append(allErrs, s.validateVolumeExpansionSupported(ctx, apiReader)...)
```

Comment:
I added an if so that API calls (access to cache) are not made if volume expansion is not done, but in fact validate is performed on all pvc's.

https://github.com/cybozu-go/moco/blob/acfa8d0e37a3178ca781635133d6249392104d84/api/v1beta2/mysqlcluster_types.go#L296-L312

Reply:
I fixed unnecessary logic:
https://github.com/cybozu-go/moco/pull/420/commits/ae635e01f203abad17fe252fb7f4a04f7b25279e

Reply:
e.g.

```diff
  volumeClaimTemplates:
    - metadata:
        name: mysql-data
      spec:
        storageClassName: not-allow-volume-expansion
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
-           storage: 500Mi
+           storage: 600Mi
    - metadata:
        name: sidecar-data
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
-           storage: 500Mi
+           storage: 600Mi
```

```
The MySQLCluster "cluster" is invalid:
* spec.volumeClaimTemplates[0].spec.storageClassName: Forbidden: storage class "not-allow-volume-expansion" is not allowed to expand volume
* spec.volumeClaimTemplates[1]: Forbidden: default storage class "standard" is not allowed to expand volume
```

Reply:
@masa213f 

oops, Sorry, the sample I presented has both storage extensions disabled to verify that multiple PVCs are validates.

> ```diff
> - storageClassName: standard-allow-volume-expansion
> + storageClassName: not-allow-volume-expansion
> ```

So in the case of the manifests you presented, I think it is correct that only `volumeClaimTemplates[1]` will get a validate error.

Reply:
Thanks for the bug report 🙏 
I will check it.

Reply:
Fixed and add tests:
https://github.com/cybozu-go/moco/pull/420/commits/41aedb6ff3b1afb079d4534c91d2ba9be64b6f7a